### PR TITLE
Implementing Ticket-Fees API

### LINKED ISSUE #3826: Implement TicketFees API with permissions
API for model `TicketFees`.

Endpoints : 
**To get all the ticket-fees** : `/v1/ticket-fees`
**To get a ticket-fees by id** : `/v1/ticket-fees/<int:id>`

Parent [issue#3640](https://github.com/fossasia/open-event-orga-server/issues/3640)

### PR DESCRIPTION
**Pull request by LuD1161**
#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

#### Short description of what this resolves:
Creating the **ticket-fees** api on the model **TicketFees**

#### Changes proposed in this pull request:
Endpoints added 
- `/ticket-fees`
- `/ticket-fees/<int:id>`

Fixes #3826

### CODE REVIEW

File: app/api/ticket_fees.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        Meta class for image_size Api Schema
+        """
+        type_ = 'ticket-fees'
+        self_view = 'v1.ticket_fee_detail'
+        self_view_kwargs = {'id': '<id>'}
+        inflect = dasherize
+
+    id = fields.Integer(dump_only=True)
+    currency = fields.Str()
+    service_fee = fields.Float()
```
Comment (poush) [inline app/api/ticket_fees.py]: I think validation can be added here to ensure amount > 0

File: app/api/ticket_fees.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    maximum_fee = fields.Float()
```
Comment (poush) [inline app/api/ticket_fees.py]: SImilary here also

**Comment by niranjan94** [comment]
Ticket Fees perms:
Admin - Read, create, edit
Other - No access

**Comment by LuD1161** [comment]
@poush @niranjan94 @SaptakS 
Sorry for the late response , I was AFK.

Can there be any checks for currency ?

File: app/api/ticket_fees.py (hunk 1 above)
Comment (LuD1161) [inline app/api/ticket_fees.py]: Done :+1:

File: app/api/ticket_fees.py (hunk 2 above)
Comment (LuD1161) [inline app/api/ticket_fees.py]: Done :+1:

**Comment by poush** [comment]
I think there should be some list of accepted currencies. @SaptakS @niranjan94 Can you please confirm ?

**Comment by niranjan94** [comment]
@poush all `ISO 4217` should be accepted

**Comment by LuD1161** [comment]
@niranjan94
Will do it by tonight.

**Comment by niranjan94** [comment]
@LuD1161 can you please not open new PRs when two of your old one's are still pending and need changes ?

**Comment by LuD1161** [comment]
Sorry for that.
I will make sure of it.

**Comment by LuD1161** [comment]
@poush @niranjan94 
Have added the list of currencies . 
Added a new list of currency so that it is directly available instead of performing operations each time on `PAYMENT_CURRENCIES`  **set** to get the list of `PAYMENT_CURRENCY_CHOICES`.

**Comment by LuD1161** [comment]
@poush @niranjan94 
Can you review it ?

File: docs/api/api_blueprint.apib [hunk 3]
```diff
@@ -5937,6 +5937,202 @@ Delete a single image size.
...
+
+Ticket Fees basically stores the fee on a ticket
+
+| Parameter | Description | Type | Required |
+|:----------|-------------|------|----------|
+| `currency`  | Currency of the ticket | string | - |
+| `service_fee` | Service fee the ticket  | float | - |
+| `maximum_fee` | Maximum fee the ticket  | float | - |
+
+## Image Size Collection [/v1/ticket-fees{?page%5bsize%5d,page%5bnumber%5d,sort,filter}]
```
Comment (magdalenesuo) [inline docs/api/api_blueprint.apib]: Correct Image Size to Ticket fees.
Please place this ticket fees section after tickets to keep the arrangement proper.
And as per [#3849(comment)](https://github.com/fossasia/open-event-orga-server/pull/3849#issuecomment-310365677), start adding the test to hook_main.py everytime you add in a new api in api doc

**Comment by SaptakS** [comment]
Please resolve conflicts and travis build.

File: docs/api/api_blueprint.apib (hunk 3 above)
Comment (LuD1161) [inline docs/api/api_blueprint.apib]: @magdalenesuo 
sorry for the late reply.
I have made the changes requested.

**Comment by LuD1161** [comment]
@SaptakS can you review this now ?

File: docs/api/api_blueprint.apib [hunk 4]
```diff
@@ -4894,6 +4894,202 @@ Delete a single ticket.
...
+                        "service-fee":"3.4",
+                        "maximum-fee":"5.4"
+                    }
+                }
+            }
+
++ Response 201 (application/vnd.api+json)
+
+        {
+          "data": "data"
```
Comment (SaptakS) [inline docs/api/api_blueprint.apib]: Don't write just "data": "data". Please write the entire response.

**Comment by LuD1161** [comment]
@SaptakS 
Sorry for the delay , I was travelling.
Can you review it now ?

**Comment by LuD1161** [comment]
@magdalenesuo
Can you review ?

File: app/api/ticket_fees.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    currency = fields.Str(validate=validate.OneOf(choices=PAYMENT_CURRENCY_CHOICES))
+    service_fee = fields.Float(validate=lambda n: n >= 0)
+    maximum_fee = fields.Float(validate=lambda n: n >= 0)
+
+
+class TicketFeeList(ResourceList):
+    """
+    List and create TicketFees
+    """
+    decorators = (is_admin,)
```
Comment (SaptakS) [inline app/api/ticket_fees.py]: Can you please use permission manager instead?

File: app/api/ticket_fees.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    schema = TicketFeesSchema
+    data_layer = {'session': db.session,
+                  'model': TicketFees}
+
+
+class TicketFeeDetail(ResourceDetail):
+    """
+    ticket_fee detail by id
+    """
+    decorators = (is_admin,)
```
Comment (SaptakS) [inline app/api/ticket_fees.py]: Same here.

File: docs/api/api_blueprint.apib [hunk 7]
```diff
@@ -4836,6 +4836,219 @@ Delete a single ticket.
...
+
+        {
+            "meta": {
+                "count": 1
+            },
+            "data": [
+                {
+                    "attributes": {
+                        "currency": "USD",
+                        "service-fee": 3.4,
```
Comment (SaptakS) [inline docs/api/api_blueprint.apib]: There should always be two decimal places when currency is involved.

File: tests/hook_main.py [hunk 8]
```diff
@@ -1361,3 +1361,55 @@ def event_sub_topic_delete(transaction):
...
+
+
+@hooks.before("Ticket Fees > Ticket Fees Collection > List Ticket Fees")
+def ticket_fees_get_list(transaction):
+    """
+    GET /ticket-fees
+    :param transaction:
+    :return:
+    """
+    transaction['skip'] = True
```
Comment (SaptakS) [inline tests/hook_main.py]: Can you please implement the Dredd testing as well? Or do you plan to do that as a separate PR?

Comment (LuD1161) [inline tests/hook_main.py]: I will do it in a separate PR.

File: app/api/ticket_fees.py (hunk 5 above)
Comment (LuD1161) [inline app/api/ticket_fees.py]: @SaptakS 
Done.

File: app/api/ticket_fees.py (hunk 6 above)
Comment (LuD1161) [inline app/api/ticket_fees.py]: @SaptakS 
Done.

File: docs/api/api_blueprint.apib (hunk 7 above)
Comment (LuD1161) [inline docs/api/api_blueprint.apib]: @SaptakS 
Done.
